File: C/test_munchausen_number.c

#include <sys/types.h>
#include <time.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int raisedto(int n);

int main(int argc, char* argv[]) {
    int n, i, number;
    int power_of_digits[10];
    clock_t start, finish;

    srand(time(NULL));

    // Find the four Munchausen numbers
    start = clock();

    for (i=0; i<10; i++) {
        power_of_digits[i] = raisedto(i);
        // printf("%d %d \n", i, power_of_digits[i]);
    }

    n = 0;
    i = 0;

    while(n < 4) {
        int sum = 0;
        for (number = i; number > 0; number /= 10) {
            int digit = number % 10;
            // find the sum of the digits raised to themselves 
            sum += power_of_digits[digit];
        }
        if (sum == i) {
            n++;
            // the sum is equal to the number itself; 
            // thus it is a munchausen number
            printf("Munchausen number %i: %i\n", n, i);
        } 
        i++;
    }
    finish = clock();
    printf("Munchausen numbers: %lf s \n", (double) (finish - start)/CLOCKS_PER_SEC);
    
    return 0;
}

int raisedto(int n) {
    if(n == 0) {
        return(0);
    }
        else {
        return((int) pow(n, n));
    }
}

File: Rust/test_munchausen_number.rs


use std::time::Instant;

fn main() {
    // Array to cache the precomputed powers of digits
    let mut power_of_digits = [0; 10];

    // Start the timer
    let start = Instant::now();

    // Precompute the digit powers
    for i in 0..10 {
        power_of_digits[i] = raisedto(i as i32);
    }

    let mut n = 0;
    let mut i = 0;

    // Find the four Munchausen numbers
    while n < 4 {
        let mut sum = 0;
        let mut number = i;

        // Extract digits and sum their precomputed powers
        while number > 0 {
            let digit = (number % 10) as usize;
            sum += power_of_digits[digit];
            number /= 10;
        }

        // Check if the sum is equal to the number itself
        if sum == i {
            n += 1;
            println!("Munchausen number {}: {}", n, i);
        }

        i += 1;
    }

    // End the timer
    let duration = start.elapsed();
    println!("Munchausen numbers: {:.6} s", duration.as_secs_f64());
}

/// Calculates n^n, with the special convention that 0^0 = 0
fn raisedto(n: i32) -> i32 {
    if n == 0 {
        0
    } else {
        // Rust's integer pow takes a u32 for the exponent
        n.pow(n as u32)
    }
}

